packages/contracts/test/plugin/plugin-setup-processor.ts

```diff
@@ -0,0 +1,818 @@
+    await managingDao.grant(
+      aragonPluginRegistry.address,
+      pluginRepoFactory.address,
+      PLUGIN_REGISTER_PERMISSION_ID
+    );
+
+    // PluginSetupProcessor
+    const PluginSetupProcessor = await ethers.getContractFactory(
+      'PluginSetupProcessor'
+    );
+    psp = await PluginSetupProcessor.deploy(
+      managingDao.address,
+      aragonPluginRegistry.address
+    );
+  });
+
+  beforeEach(async function () {
+    // Target DAO to be used as an example DAO
+    targetDao = await deployNewDAO(ownerAddress);
+
+    // Create and register a plugin on the AragonPluginRegistry
+    const tx = await pluginRepoFactory.createPluginRepoWithVersion(
+      'PluginSetupV1Mock',
+      [1, 0, 0],
+      pluginSetupV1Mock.address,
+      '0x00',
+      ownerAddress
+    );
+
+    const event = await decodeEvent(tx, EVENTS.PluginRepoRegistered);
+
+    pluginSetupMockRepoAddress = event.args.pluginRepo;
+
+    // Add PluginSetupV2Mock to the PluginRepo.
+    const PluginRepo = await ethers.getContractFactory('PluginRepo');
+    pluginRepo = PluginRepo.attach(pluginSetupMockRepoAddress);
+    await pluginRepo.createVersion(
+      [2, 0, 0],
+      pluginSetupV2Mock.address,
+      '0x00'
+    );
+
+    // Grant
+    await targetDao.grant(targetDao.address, psp.address, ROOT_PERMISSION_ID);
+  });
+
+  describe('Install Plugin', function () {
+    beforeEach(async () => {
+      // Grant necessary permission to `ownerAddress` so it can install plugin on behalf of the DAO,
+      // this way we don't have to create DAO execute call.
+      await targetDao.grant(
+        psp.address,
+        ownerAddress,
+        PROCESS_INSTALL_PERMISSION_ID
+      );
+    });
+
+    it('PrepareInstallation: reverts if `PluginSetupRepo` do not exist on `AragonPluginRegistry`', async () => {
+      const data = '0x';
+      const pluginSetupRepoAddr = ADDRESS_TWO;
+
+      await expect(
+        psp.prepareInstallation(
+          targetDao.address,
+          pluginSetupV1Mock.address,
+          pluginSetupRepoAddr,
+          data
+        )
+      ).to.be.revertedWith(customError('EmptyPluginRepo'));
+    });
+
+    it('PrepareInstallation: retrun correctly the permissions', async () => {
+      const {plugin, helpers, prepareInstallpermissions} =
+        await prepareInstallation(
+          psp,
+          targetDao.address,
+          pluginSetupV1Mock.address,
+          pluginSetupMockRepoAddress,
+          EMPTY_DATA
+        );
+
+      expect(plugin).not.to.be.equal(AddressZero);
+      expect(helpers.length).to.be.equal(1);
+      expect(prepareInstallpermissions).to.deep.equal([
+        [
+          Op.Grant,
+          targetDao.address,
+          plugin,
+          AddressZero,
+          ethers.utils.id('EXECUTE_PERMISSION'),
+        ],
+        [
+          Op.Grant,
+          plugin,
+          helpers[0],
+          AddressZero,
+          ethers.utils.id('SETTINGS_PERMISSION'),
+        ],
+      ]);
+    });
```

Comment:
You can nest `describe` statements to give the tests more structure. See, for example:
```suggestion
  describe('Install Plugin', function () {
    beforeEach(async () => {
      // Grant necessary permission to `ownerAddress` so it can install plugin on behalf of the DAO,
      // this way we don't have to create DAO execute call.
      await targetDao.grant(
        psp.address,
        ownerAddress,
        PROCESS_INSTALL_PERMISSION_ID
      );
    });

    describe('PrepareInstallation', function () {
      it('Reverts if `PluginSetupRepo` do not exist on `AragonPluginRegistry`', async () => {
        const data = '0x';
        const pluginSetupRepoAddr = ADDRESS_TWO;

        await expect(
          psp.prepareInstallation(
            targetDao.address,
            pluginSetupV1Mock.address,
            pluginSetupRepoAddr,
            data
          )
        ).to.be.revertedWith(customError('EmptyPluginRepo'));
      });

      it('Return correctly the permissions', async () => {
        const {plugin, helpers, prepareInstallpermissions} =
          await prepareInstallation(
            psp,
            targetDao.address,
            pluginSetupV1Mock.address,
            pluginSetupMockRepoAddress,
            EMPTY_DATA
          );

        expect(plugin).not.to.be.equal(AddressZero);
        expect(helpers.length).to.be.equal(1);
        expect(prepareInstallpermissions).to.deep.equal([
          [
            Op.Grant,
            targetDao.address,
            plugin,
            AddressZero,
            ethers.utils.id('EXECUTE_PERMISSION'),
          ],
          [
            Op.Grant,
            plugin,
            helpers[0],
            AddressZero,
            ethers.utils.id('SETTINGS_PERMISSION'),
          ],
        ]);
      });
    });
```

Reply:
amended